Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,55 +49,162 @@ void RebinToWorkspace::init() {
                   "Workspace2D histogram.");
 }
 
+namespace {
+bool needToRebin(const MatrixWorkspace_sptr &left,
+                 const MatrixWorkspace_sptr &rght) {
+  // if pointers match they are the same object
+  if (left == rght)
+    return false;
+
+  // see if there is the same number of histograms
+  const size_t numHist = left->getNumberHistograms();
+  if (numHist != rght->getNumberHistograms())
+    return true;
+
+  // look for first non-equal x-axis between the workspaces
+  for (size_t i = 0; i < numHist; ++i) {
+    if (left->getSpectrum(i).x() != rght->getSpectrum(i).x()) {
+      return true;
+    }
+  }
+  // everything must be the same
+  return false;
+}
+} // namespace
+
 /**
  * Execute the algorithm
  */
 void RebinToWorkspace::exec() {
   // The input workspaces ...
   MatrixWorkspace_sptr toRebin = getProperty("WorkspaceToRebin");
   MatrixWorkspace_sptr toMatch = getProperty("WorkspaceToMatch");
-  bool PreserveEvents = getProperty("PreserveEvents");
-
-  // First we need to create the parameter vector from the workspace with which
-  // we are matching
-  std::vector<double> rb_params = createRebinParameters(toMatch);
-
-  IAlgorithm_sptr runRebin = createChildAlgorithm("Rebin");
-  runRebin->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
-  runRebin->setPropertyValue("OutputWorkspace", "rebin_out");
-  runRebin->setProperty("params", rb_params);
-  runRebin->setProperty("PreserveEvents", PreserveEvents);
-  runRebin->setProperty("IgnoreBinErrors", true);
-  runRebin->executeAsChildAlg();
-  progress(1);
-  MatrixWorkspace_sptr ws = runRebin->getProperty("OutputWorkspace");
-  setProperty("OutputWorkspace", ws);
+  m_preserveEvents = getProperty("PreserveEvents");
+  m_isEvents = bool(boost::dynamic_pointer_cast<const EventWorkspace>(toRebin));
+
+  if (needToRebin(toRebin, toMatch)) {
+    g_log.information("Rebinning");
+    if (m_isEvents && (!m_preserveEvents)) {
+      this->histogram(toRebin, toMatch);
+    } else {
+      this->rebin(toRebin, toMatch);
+    }
+  } else { // don't need to rebin
+    g_log.information(
+        "WorkspaceToRebin and WorkspaceToMatch already have matched binning");
+    auto outputWS = this->createOutputWorkspace(toRebin, 1.);
```

Comment:
This control flow is complicated, you have implemented two ways to do the same thing, I think you can simplify as follows:

1. `needToRebin` should always return `true` `if(m_isEvents && (!m_preserveEvents))`.
2. You can then completely remove the call to `ConvertToMatrixWorkspace` in `createOutputWorkspace`.
3. Then you will notice that you can move the check for `needToRebin` into `rebin`, after the place where `createOutputWorkspace` is called, and remove the whole `if` statement here.

Reply:
Good call. I missed that.